### ISSUE TITLE
NDArrayMathCPU still tries to use GPU?

### ISSUE THREAD

**Issue by silky** [tensorflow/tfjs-core#116]
maybe it's my lack of understanding of this library, but i've got the following error

```
Uncaught Error: GPU not intialized.
    at throwIfGPUNotInitialized (deeplearn-latest.js:3211)
    at Array2D.NDArray.getValues (deeplearn-latest.js:3347)
    at Array2D.get (deeplearn-latest.js:3506)
    at normalGetter (deeplearn-latest.js:2357)
    at NDArrayMathCPU.matMulInternal (deeplearn-latest.js:2374)
    at deeplearn-latest.js:1803
    at NDArrayMathCPU.NDArrayMath.executeOp (deeplearn-latest.js:1810)
    at NDArrayMathCPU.NDArrayMath.matMul (deeplearn-latest.js:1803)
    at deeplearn-latest.js:6178
    at NDArrayMathCPU.NDArrayMath.scope (deeplearn-latest.js:1698)
```

this is in when i instantiate math like:

```
var math = new deeplearn.NDArrayMathCPU();
```

if i just use "MathGPU" it works fine.

i'm on a lenovo ubuntu laptop and i only have intel onboard graphics, so it's a bit odd.

oddly, i only get this error when i do another forward pass on my network in the `requestAnimationFrame` function; the first forward pass works.

my forward call looks something like:

```
	requestAnimationFrame( function () {
		math.scope(function (){
			forward( net, session, math, ... );
		});
	});
```

**Comment by nsthorat** [comment]
Oh that is really weird, and not what's supposed to happen.

Can you show me the entirety of your code?

**Comment by silky** [comment]
it's here: https://github.com/silky/simple-cppn-deeplearn.js/blob/master/cppn.js

if you uncomment https://github.com/silky/simple-cppn-deeplearn.js/blob/master/cppn.js#L138 and switch https://github.com/silky/simple-cppn-deeplearn.js/blob/master/cppn.js#L155 you should see the problem.

**Comment by arthurjdam** [comment]
Ah that's correct; there's a bunch of GPGPU references in NDArray. I guess these should all live in NDArrayMathGPU since it should be stateless (i.e. can run either on GPU or CPU)?

**Comment by arthurjdam** [comment]
@silky I did some digging. It doesn't try to actually call the GPU, it's a test which checks if data.values is undefined, in which case it assumes the math is performed via WebGL (which is obviously wrong by itself, but that's a separate issue all-together).

Either way, I've updated your app to make a new math object per function; for some reason NDArrayMathCPU changes its state (not yet sure why, but this seems a little closer to expected behavior, since there's a ton of calls to `this`, you should re-instantiate the object instead of serve it globally). It's unusably slow, but at least solves this issue :) 

https://github.com/arthurjdam/simple-cppn-deeplearn.js/tree/feature-cpu

**Comment by nsthorat** [comment]
@silky sorry for the late response, things have been a little hectic.

The issue is that your NDArray is disposed, and you're trying to call getValues on it. We have a bad error message here, I will fix that. We basically assume if values is null it's on the GPU, this is a wrong assumption because it could be a disposed CPU NDArray.

I'm late to responding so I don't see the lines to uncomment, but you should check your scopes. math.scope() cleans up any outputs of math operations that are used inside of that function closure, so if what you're calling getValues on lives outside that scope, it will be disposed!

If you can put up a version of the demo on your github pages that shows the error, I can step through it live.

On another note, your CPPN is slow because of this:
https://github.com/silky/simple-cppn-deeplearn.js/blob/master/cppn.js#L123

Basically, every frame you're downloading to the CPU (getValues() is a blocking readPixels call), and then constructing an ImageData object that backs the canvas. This is super slow.

What you need to do make this faster (and not round trip to the CPU) is construct the GPGPUContext from the canvas you want to draw to:
https://github.com/PAIR-code/deeplearnjs/blob/master/demos/nn-art/cppn.ts#L75

And use this shader:
https://github.com/PAIR-code/deeplearnjs/blob/master/demos/nn-art/cppn.ts#L173

**Comment by dsmilkov** [comment]
Thanks everyone for a great discussion. Closing since this is a bit outdated. Feel free to reopen with a discussion that reflects the latest state of the library.
